### PR TITLE
Fix issue psalm querybuilder

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: src/Query/QueryBuilder.php [hunk 1]
```diff
@@ -1870,32 +1881,9 @@ protected function extractAlias(string $table)
         return false;
     }
 
-    public function buildWith($withs, &$params): string
```
Comment (samdark) [inline src/Query/QueryBuilder.php:1873]: Why is it removed?

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/Query/QueryBuilder.php (hunk 1 above)
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1873]: It was replaced by buildWithQueries (), both methods are the same, and you can check that it is not used.

File: src/Query/QueryBuilder.php [hunk 2]
```diff
@@ -1663,8 +1670,8 @@ public function buildOrderBy(array $columns, array &$params = []): string
     }
 
     /**
-     * @param int|object|null $limit
-     * @param int|object|null $offset
+     * @param Expression|int|null $limit
+     * @param Expression|int|null $offset
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1674]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 3]
```diff
@@ -1663,8 +1670,8 @@ public function buildOrderBy(array $columns, array &$params = []): string
     }
 
     /**
-     * @param int|object|null $limit
-     * @param int|object|null $offset
+     * @param Expression|int|null $limit
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1673]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 4]
```diff
@@ -1651,7 +1658,7 @@ public function buildOrderBy(array $columns, array &$params = []): string
         $orders = [];
 
         foreach ($columns as $name => $direction) {
-            if ($direction instanceof ExpressionInterface) {
+            if ($direction instanceof Expression) {
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1661]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 5]
```diff
@@ -1636,6 +1641,8 @@ public function buildOrderByAndLimit(
 
     /**
      * @param array $columns
+     * @psalm-param array<string, Expression|int|string> $columns
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1644]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 6]
```diff
@@ -1563,6 +1563,8 @@ public function buildWhere($condition, array &$params = []): string
 
     /**
      * @param array $columns
+     * @psalm-param array<string, Expression|string> $columns
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1566]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 7]
```diff
@@ -1607,8 +1610,10 @@ public function buildHaving($condition, array &$params = []): string
      * @param string $sql the existing SQL (without ORDER BY/LIMIT/OFFSET).
      * @param array $orderBy the order by columns. See {@see Query::orderBy} for more details on how to specify this
      * parameter.
-     * @param int|object|null $limit the limit number. See {@see Query::limit} for more details.
-     * @param int|object|null $offset the offset number. See {@see Query::offset} for more details.
+     * @psalm-param array<string, Expression|int|string> $orderBy
+     *
+     * @param Expression|int|null $limit the limit number. See {@see Query::limit} for more details.
+     * @param Expression|int|null $offset the offset number. See {@see Query::offset} for more details.
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1616]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 8]
```diff
@@ -1607,8 +1610,10 @@ public function buildHaving($condition, array &$params = []): string
      * @param string $sql the existing SQL (without ORDER BY/LIMIT/OFFSET).
      * @param array $orderBy the order by columns. See {@see Query::orderBy} for more details on how to specify this
      * parameter.
-     * @param int|object|null $limit the limit number. See {@see Query::limit} for more details.
-     * @param int|object|null $offset the offset number. See {@see Query::offset} for more details.
+     * @psalm-param array<string, Expression|int|string> $orderBy
+     *
+     * @param Expression|int|null $limit the limit number. See {@see Query::limit} for more details.
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1615]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 9]
```diff
@@ -1574,8 +1576,9 @@ public function buildGroupBy(array $columns, array &$params = []): string
         if (empty($columns)) {
             return '';
         }
+
         foreach ($columns as $i => $column) {
-            if ($column instanceof ExpressionInterface) {
+            if ($column instanceof Expression) {
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:1581]: Only `Expression::class` is used.

File: src/Query/QueryBuilder.php [hunk 10]
```diff
@@ -105,8 +106,8 @@ class QueryBuilder
      *
      * In case you want to add custom conditions support, use the {@see setConditionClasses()} method.
      *
-     * @see setConditonClasses()
-     * @see defaultConditionClasses()
+     * {@see setConditonClasses()}
```
Comment (samdark) [inline src/Query/QueryBuilder.php:109]: No need for `{` and `}`. These are needed when you use it inside of text.

File: src/Query/QueryBuilder.php [hunk 11]
```diff
@@ -88,6 +87,8 @@ class QueryBuilder
      * @var array the abstract column types mapped to physical column types.
      * This is mainly used to support creating/modifying tables using DB-independent data type specifications.
      * Child classes should override this property to declare supported type mappings.
+     *
+     * @psalm-var array<array-key, string>
```
Comment (samdark) [inline src/Query/QueryBuilder.php:91]:
```suggestion
     * @psalm-var array<string, string>
```

**Review by samdark — APPROVED** [review]

File: src/Query/QueryBuilder.php (hunk 10 above)
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php:109]: done.

File: src/Query/QueryBuilder.php (hunk 11 above)
Comment (terabytesoftw) [reply src/Query/QueryBuilder.php]: Done.

**Pull request merged**